### PR TITLE
Auto refresh Cassandra nodes

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Building on https://github.com/palantir/atlasdb/pull/3599 (but not on a fork). 

**Implementation Description (bullets)**:
Other than some light refactoring to make it easier to parse, same as @amandachow 's

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added tests for the relevant behaviour that mock out CassandraService. This is probably enough?

**Concerns (what feedback would you like?)**:
We need to follow up, either here or in a separate PR, on allowing the config servers to be live reloadable.
 
**Priority (whenever / two weeks / yesterday)**:
This week

### CODE REVIEW

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java [hunk 1]
```diff
@@ -121,7 +138,8 @@ public void shouldRetryOnSameNodeToFailureAndThenRedirect() {
 
         CassandraClientPoolImpl cassandraClientPool = throwingClientPoolWithServersInCurrentPool(
                 ImmutableSet.copyOf(hostList), new SocketTimeoutException());
-        runNoopOnHostWithRetryWithException(hostList.get(0), cassandraClientPool);
+        assertThatThrownBy(() -> runNoopWithRetryOnHost(hostList.get(0), cassandraClientPool))
+                .isInstanceOf(Exception.class);
```
Comment (sandorw) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java]: Worth doing anything more here?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java [hunk 2]
```diff
@@ -230,6 +248,113 @@ public void attemptsShouldBeCountedPerHost() {
...
+        createClientPool();
+        assertThat(poolServers).containsExactlyInAnyOrder(HOST_1, HOST_2);
+
+        setCassandraServersTo(HOST_1, HOST_2, HOST_3);
+        deterministicExecutor.tick(config.poolRefreshIntervalSeconds(), TimeUnit.SECONDS);
+        assertThat(poolServers).containsExactlyInAnyOrder(HOST_1, HOST_2);
+    }
+
+    @Test
+    public void hostsAreResetToCondigOnRefreshWhenRefreshIsDisabled() {
```
Comment (sandorw) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java]: nit: Config

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 3]
```diff
@@ -132,32 +129,32 @@ static CassandraClientPoolImpl createImplForTest(
         return cassandraClientPool;
     }
 
-    public static CassandraClientPool create(MetricsManager metricsManager, CassandraKeyValueServiceConfig config) {
-        return create(metricsManager,
+    @VisibleForTesting
+    static CassandraClientPoolImpl createImplForTest(
```
Comment (sandorw) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: more test factories 😢 😢

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 4]
```diff
@@ -254,43 +275,38 @@ int getMaxTriesTotal() {
...
         blacklist.checkAndUpdate(cassandra.getPools());
 
-        Set<InetSocketAddress> serversToAdd = Sets.newHashSet(config.servers());
-        Set<InetSocketAddress> serversToRemove = ImmutableSet.of();
-
         if (config.autoRefreshNodes()) {
-            serversToAdd.addAll(cassandra.refreshTokenRanges());
+            setServersInPoolTo(cassandra.refreshTokenRangesAndGetServers());
+        } else {
+            setServersInPoolTo(config.servers());
```
Comment (sandorw) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: Is there any case where auto refresh should be off? Not in scope for this PR, but should we support it?

Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: I can't imagine why we'd want that, but maybe cassops people can think of a scenario

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java (hunk 3 above)
Comment (gmaretic) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: I know, it's horrible :( Maybe we should add this to the list of things that could do with a refactoring?

**Comment by sandorw** [comment]
New tests were nice and clear by the way 😄

**Pull request merged**